deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/RecurrentAttentionLayer.java

```diff
@@ -0,0 +1,326 @@
+        if (inputType == null || inputType.getType() != InputType.Type.RNN) {
+            throw new IllegalStateException("Invalid input for Recurrent Attention layer (layer index = " + layerIndex
+                    + ", layer name = \"" + getLayerName() + "\"): expect RNN input type with size > 0. Got: "
+                    + inputType);
+        }
+
+        InputType.InputTypeRecurrent itr = (InputType.InputTypeRecurrent) inputType;
+
+
+        return InputType.recurrent(nOut, itr.getTimeSeriesLength());
+    }
+
+    @Override
+    public void defineParameters(SDLayerParams params) {
+        params.clear();
+
+        params.addWeightParam(WEIGHT_KEY, nIn, nOut);
+        params.addWeightParam(RECURRENT_WEIGHT_KEY, nOut, nOut);
+        if(hasBias){
+            params.addBiasParam(BIAS_KEY, nOut);
+        }
+
+        if(projectInput){
+            params.addWeightParam(WEIGHT_KEY_QUERY_PROJECTION, nHeads, headSize, nOut);
+            params.addWeightParam(WEIGHT_KEY_KEY_PROJECTION,   nHeads, headSize, nIn);
+            params.addWeightParam(WEIGHT_KEY_VALUE_PROJECTION, nHeads, headSize, nIn);
+            params.addWeightParam(WEIGHT_KEY_OUT_PROJECTION, nHeads * headSize, nOut);
+        }
+    }
+
+    @Override
+    public void initializeParameters(Map<String, INDArray> params) {
+        try (MemoryWorkspace ws = Nd4j.getWorkspaceManager().scopeOutOfWorkspaces()) {
+            for (Map.Entry<String, INDArray> e : params.entrySet()) {
+                final String keyName = e.getKey();
+                switch (keyName) {
+                    case WEIGHT_KEY:
+                        WeightInitUtil.initWeights(nIn, nOut, e.getValue().shape(), weightInit, null, 'c', e.getValue());
+                        break;
+                    case RECURRENT_WEIGHT_KEY:
+                        WeightInitUtil.initWeights(nOut, nOut, e.getValue().shape(), weightInit, null, 'c', e.getValue());
+                        break;
+                    case BIAS_KEY:
+                        e.getValue().assign(0);
+                        break;
+                    case WEIGHT_KEY_OUT_PROJECTION:
+                        WeightInitUtil.initWeights(nIn, headSize, e.getValue().shape(), weightInit, null, 'c', e.getValue());
+                        break;
+                    default:
+                        WeightInitUtil.initWeights(nHeads * headSize, nOut, e.getValue().shape(), weightInit, null, 'c', e.getValue());
+                        break;
+                }
+            }
+        }
+    }
+
+    @Override
+    public void applyGlobalConfigToLayer(NeuralNetConfiguration.Builder globalConfig) {
+        if (activation == null) {
+            activation = SameDiffLayerUtils.fromIActivation(globalConfig.getActivationFn());
+        }
+    }
+
+    @Override
+    public SDVariable defineLayer(SameDiff sameDiff, SDVariable layerInput, Map<String, SDVariable> paramTable, SDVariable mask) {
```

Comment:
This implementation will only work with sequences that are identical length for all minibatches - but we don't have any checks for that anywhere as far as I can see... let's add checks and also test that. Let's also open an issue about it (once merged), and switch this to proper SameDiff loops when they are ready.

Also not sure how safe/reliable the ```final long[] inputShape = layerInput.getShape();  final val timeSteps = inputShape[2];``` is, though I guess if tests are OK...

Reply:
About the same time that you've commented, I noticed that as an issue as well. Not really sure how to check for different sequence length here though, as `defineLayer` is always only called once.

Reply:
Two ways I can think of doing this...
One is we solve the general problem here: https://github.com/deeplearning4j/deeplearning4j/issues/7527
Though that might take too long.

Another is we add some sort of "validateInput(INDArray[])" method on the DL4J samediff layers, that defaults to no-op (but can be overridden if required). Then call that just before forward pass. That should be straightforward to add I think.

Reply:
I think for now we can go with the second option. I'll add it to the SameDiffLayer and SameDiffVertex base classes